src/stats/kde.nim

```diff
@@ -0,0 +1,209 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ../tensor/tensor,
+       ./distributions
+
+import std / [math, strutils]
+when (NimMajor, NimMinor, NimPatch) >= (1, 3, 0):
+  export nimIdentNormalize # for parseEnum
+
+type
+  KernelKind* = enum
+    knCustom = "custom"
+    knBox = "box"
+    knTriangular = "triangular"
+    knTrig = "trigonometric"
+    knEpanechnikov = "epanechnikov"
+    knGauss = "gauss"
+
+  KernelFunc* = proc(x, x_i, bw: float): float {.inline.}
+
+template makeKernel*(fn: untyped): untyped =
+  proc `fn Kernel`*(x, x_i, bw: float): float {.inline.} =
+    let val = (x - x_i) / bw
+    result = fn(val)
+
+makeKernel(box)
+makeKernel(triangular)
+makeKernel(trigonometric)
+makeKernel(epanechnikov)
+
+# manually build the gauss kerne, since the underlying distribution relies
+# on more than 1 value.
+proc gaussKernel*(x, x_i, bw: float): float {.inline.} =
+  gauss(x = x, mean = x_i, sigma = bw, norm = true)
+
+proc getCutoff(bw: float, kind: KernelKind): float =
+  ## calculates a reasonable cutoff for a given `KernelKind` for a bandwidth `bw`
+  case kind
+  of knBox: result = 0.5 * bw
+  of knTriangular: result = bw
+  of knTrig: result = 0.5 * bw
+  of knEpanechnikov: result = bw
+  of knGauss: result = 3.0 * bw # 3 sigma amounts to 99.7% of gaussian kernel contribution
+  of knCustom: doAssert false, "`getCutoff` must not be called with custom kernel!"
+
+proc getKernelFunc(kind: KernelKind): KernelFunc =
+  case kind
+  of knBox: result = boxKernel
+  of knTriangular: result = triangularKernel
+  of knTrig: result = trigonometricKernel
+  of knEpanechnikov: result = epanechnikovKernel
+  of knGauss: result = gaussKernel
+  of knCustom: doAssert false, "`getKernelFunc` must not be called with custom kernel!"
+
+proc findWindow[T](dist: T, s: T, t: Tensor[T], oldStart = 0, oldStop = 0): (int, int) =
+  ## returns the index (start or stop) given a distance `dist` that
+  ## is allowed between s - x[j]
+  # `s` and `x` must be sorted
+  var
+    startFound = false
+    stopFound = false
+  var j = oldStart
+  while j < t.size:
+    if not startFound and abs(s - t[j]) < dist:
+      startFound = true
+      result[0] = j
+      j = if oldStop < j: j else: oldStop
+      continue
+    elif startFound and not stopFound and abs(s - t[j]) > dist:
+      stopFound = true
+      result[1] = j
+      break
+    inc j
+ # set to max, if we left the while loop naturally
+  if result[1] == 0: result[1] = t.size
+  assert result[1] > result[0]
+
+proc kde*[T: SomeNumber; U: int | Tensor[SomeNumber] | openArray[SomeNumber]](
+    t: Tensor[T],
+    kernel: KernelFunc,
```

Comment:
should this be `static` if we want `{.inline.}` to work (though I'm pretty sure static proc will crash the compiler).

Reply:
I'll give it a try.